modules/jdbc/src/main/java/org/testcontainers/jdbc/ConnectionUrl.java

```diff
@@ -0,0 +1,209 @@
+            databaseHost = Optional.of(dbInstanceMatcher.group(1));
+            databasePort = Optional.ofNullable(dbInstanceMatcher.group(3)).map(value -> Integer.valueOf(value));
+            databaseName = Optional.of(dbInstanceMatcher.group(4));
+        }
+
+        queryString = Optional.ofNullable(urlMatcher.group(5));
+        getQueryParameters();
+
+        Matcher matcher = Patterns.INITSCRIPT_MATCHING_PATTERN.matcher(this.getUrl());
+        if (matcher.matches()) {
+            initScriptPath = Optional.ofNullable(matcher.group(2));
+        }
+
+        Matcher funcMatcher = Patterns.INITFUNCTION_MATCHING_PATTERN.matcher(this.getUrl());
+        if (funcMatcher.matches()) {
+            initFunction = Optional.of(new InitFunctionDef(funcMatcher.group(2), funcMatcher.group(4)));
+        }
+
+        Matcher daemonMatcher = Patterns.DAEMON_MATCHING_PATTERN.matcher(this.getUrl());
+        inDaemonMode = daemonMatcher.matches() ? Boolean.parseBoolean(daemonMatcher.group(2)) : false;
+
+    }
+
+    /**
+     * Get the TestContainers Parameters such as Init Function, Init Script path etc.
+     *
+     * @return {@link Map}
+     */
+    public Map<String, String> getContainerParameters() {
+
+        Map<String, String> results = new HashMap<>();
+
+        Matcher matcher = Patterns.TC_PARAM_MATCHING_PATTERN.matcher(this.getUrl());
+        while (matcher.find()) {
+            String key = matcher.group(1);
+            String value = matcher.group(2);
+            results.put(key, value);
+        }
+
+        return results;
+    }
+
+    /**
+     * Get all Query paramters specified in the Connection URL after ?. This also includes TestContainers parameters.
+     *
+     * @return {@link Map}
+     */
+    public Map<String, String> getQueryParameters() {
+
+        Map<String, String> results = new HashMap<>();
+        StringJoiner query = new StringJoiner("&");
+        Matcher matcher = Patterns.QUERY_PARAM_MATCHING_PATTERN.matcher(this.getQueryString().orElse(""));
+        while (matcher.find()) {
+            String key = matcher.group(1);
+            String value = matcher.group(2);
+            if (!key.startsWith("TC_")) query.add(key + "=" + value);
+            results.put(key, value);
+        }
+
+        queryString = Optional.of("?" + query.toString());
+        return results;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (Objects.isNull(obj) || !(obj instanceof ConnectionUrl)) return false;
+        return this.getUrl().equals(((ConnectionUrl) obj).getUrl());
+    }
+
+    /**
+     * This interface defines the Regex Patterns used by {@link ConnectionUrl}.
+     *
+     * @author manikmagar
+     */
+    public interface Patterns {
+        final Pattern URL_MATCHING_PATTERN = Pattern.compile("jdbc:tc:([a-z]+)(:([^:]+))?://([^\\?]+)(\\?.*)?");
+
+        final Pattern ORACLE_URL_MATCHING_PATTERN = Pattern.compile("jdbc:tc:([a-z]+)(:([^(thin:)]+))?:thin:@([^\\?]+)(\\?.*)?");
```

Comment:
I apologise for putting you through the pain of working with these regular expressions
😂

Reply:
Actually, I liked it. I never used these many and types of regex pattern, got to learn new things ;), So Thank you!